File: src/share.ts
import { SHARE_SIL_VERSION, SITE_URL } from "./config";
import { dailyPath } from "./game/daily";
import { hashString, RNG } from "./game/rng";
import { formatScore } from "./game/scoring";
import type { Mode } from "./game/types";

const BG = "#0B0B0C";
const SURFACE = "#161618";
const TEXT = "#F4F1EA";
const MUTED = "#8A8680";
const ACCENT = "#E8FF47";

export const SHARE_SILS = [
  "circle",
  "triangle",
  "diamond",
  "bars",
  "slash",
  "chevron",
  "capsule",
  "plus",
] as const;

export type ShareSil = (typeof SHARE_SILS)[number];

export interface CardInput {
  score: number;
  roundLabel: string;
  newBest: boolean;
  mode: Mode;
  dateId: string;
}

/** Live host, e.g. temporary-zippy-mistral-mg92d6h.vercel.app — not SKEW.GAME. */
export function shareLockup(siteUrl = SITE_URL): string {
  const raw = siteUrl.replace(/\/$/, "");
  try {
    return new URL(raw).host;
  } catch {
    return raw.replace(/^https?:\/\//, "");
  }
}

export function challengeOrigin(fallbackOrigin: string): string {
  return (SITE_URL || fallbackOrigin).replace(/\/$/, "");
}

export function challengeText(score: number, origin: string, dateId: string): string {
  const link = `${challengeOrigin(origin)}${dailyPath(dateId)}`;
  return `SKEW — Find the fake. I scored ${formatScore(score)}. Can you beat me?\n${link}`;
}

export function shareSilhouetteKey(mode: Mode, dateId: string): string {
  return mode === "daily" ? `daily:${dateId}` : "endless";
}

/** Decorative trio from a share-only hash. Never the daily puzzle stream. */
export function shareSilhouettes(mode: Mode, dateId: string): [ShareSil, ShareSil, ShareSil] {
  const seed = hashString(`${SHARE_SIL_VERSION}:${shareSilhouetteKey(mode, dateId)}`);
  const rng = new RNG(seed);
  return [rng.pick(SHARE_SILS), rng.pick(SHARE_SILS), rng.pick(SHARE_SILS)];
}

export function cardDateLine(mode: Mode, dateId: string): string | null {
  if (mode !== "daily" || !dateId) return null;
  return `${dateId} UTC`;
}

export function cardMetaLine(roundLabel: string, newBest: boolean): string {
  return newBest ? `${roundLabel}  ·  New best` : roundLabel;
}

export function silhouetteSvg(kind: ShareSil): string {
  const fill = TEXT;
  switch (kind) {
    case "circle":
      return `<circle r="22" fill="${fill}" />`;
    case "triangle":
      return `<polygon points="0,-24 22,20 -22,20" fill="${fill}" />`;
    case "diamond":
      return `<polygon points="0,-24 20,0 0,24 -20,0" fill="${fill}" />`;
    case "bars":
      return `<rect x="-18" y="-20" width="8" height="40" rx="3" fill="${fill}" /><rect x="-4" y="-20" width="8" height="40" rx="3" fill="${fill}" /><rect x="10" y="-20" width="8" height="40" rx="3" fill="${fill}" />`;
    case "slash":
      return `<path d="M-16,18 L16,-18" fill="none" stroke="${fill}" stroke-width="8" stroke-linecap="round" />`;
    case "chevron":
      return `<path d="M-16,-16 L14,0 L-16,16" fill="none" stroke="${fill}" stroke-width="7" stroke-linecap="round" stroke-linejoin="round" />`;
    case "capsule":
      return `<rect x="-10" y="-22" width="20" height="44" rx="10" fill="${fill}" />`;
    case "plus":
      return `<path d="M-7,-22 H7 V-7 H22 V7 H7 V22 H-7 V7 H-22 V-7 H-7 Z" fill="${fill}" />`;
    default: {
      const _x: never = kind;
      return _x;
    }
  }
}

export function shareBoardMarkup(mode: Mode, dateId: string): string {
  return shareSilhouettes(mode, dateId)
    .map(
      (kind) =>
        `<span class="card-sil" aria-hidden="true"><svg viewBox="0 0 100 100"><g transform="translate(50 50)">${silhouetteSvg(kind)}</g></svg></span>`,
    )
    .join("");
}

export async function renderScoreCard(input: CardInput): Promise<Blob> {
  const w = 1080;
  const h = 1350;
  const canvas = document.createElement("canvas");
  canvas.width = w;
  canvas.height = h;
  const ctx = canvas.getContext("2d");
  if (!ctx) throw new Error("canvas");

  ctx.fillStyle = BG;
  ctx.fillRect(0, 0, w, h);

  ctx.fillStyle = TEXT;
  ctx.font = "700 64px ui-sans-serif, system-ui, sans-serif";
  ctx.textAlign = "center";
  setTracking(ctx, "16px");
  ctx.fillText("SKEW", w / 2, 150);

  drawBrandBars(ctx, w / 2, 230);

  const sils = shareSilhouettes(input.mode, input.dateId);
  drawSilBoard(ctx, w / 2, 520, sils);

  setTracking(ctx, "0px");
  ctx.font = "800 128px ui-sans-serif, system-ui, sans-serif";
  ctx.fillStyle = TEXT;
  ctx.fillText(formatScore(input.score), w / 2, 820);

  ctx.font = "500 34px ui-monospace, SFMono-Regular, Menlo, monospace";
  ctx.fillText(cardMetaLine(input.roundLabel, input.newBest), w / 2, 880);

  const dateLine = cardDateLine(input.mode, input.dateId);
  if (dateLine) {
    ctx.fillStyle = MUTED;
    ctx.font = "500 28px ui-monospace, SFMono-Regular, Menlo, monospace";
    ctx.fillText(dateLine, w / 2, 928);
  }

  ctx.fillStyle = MUTED;
  ctx.font = "500 34px ui-sans-serif, system-ui, sans-serif";
  ctx.fillText("Can you beat me?", w / 2, 1040);

  ctx.fillStyle = ACCENT;
  ctx.fillRect(w / 2 - 48, 1076, 96, 8);

  ctx.fillStyle = TEXT;
  ctx.font = "600 26px ui-sans-serif, system-ui, sans-serif";
  setTracking(ctx, "1px");
  ctx.fillText(shareLockup(), w / 2, 1188);

  return await new Promise((resolve, reject) => {
    canvas.toBlob((blob) => (blob ? resolve(blob) : reject(new Error("toBlob"))), "image/png");
  });
}

function drawSilBoard(
  ctx: CanvasRenderingContext2D,
  cx: number,
  cy: number,
  sils: [ShareSil, ShareSil, ShareSil],
): void {
  const cardW = 220;
  const cardH = 300;
  const gap = 28;
  const total = cardW * 3 + gap * 2;
  let x = cx - total / 2;
  sils.forEach((kind) => {
    rounded(ctx, x, cy - cardH / 2, cardW, cardH, 28);
    ctx.fillStyle = SURFACE;
    ctx.fill();
    ctx.save();
    ctx.translate(x + cardW / 2, cy);
    drawSil(ctx, kind);
    ctx.restore();
    x += cardW + gap;
  });
}

function drawSil(ctx: CanvasRenderingContext2D, kind: ShareSil): void {
  ctx.fillStyle = TEXT;
  ctx.strokeStyle = TEXT;
  ctx.lineCap = "round";
  ctx.lineJoin = "round";
  switch (kind) {
    case "circle":
      ctx.beginPath();
      ctx.arc(0, 0, 58, 0, Math.PI * 2);
      ctx.fill();
      break;
    case "triangle":
      ctx.beginPath();
      ctx.moveTo(0, -64);
      ctx.lineTo(58, 52);
      ctx.lineTo(-58, 52);
      ctx.closePath();
      ctx.fill();
      break;
    case "diamond":
      ctx.beginPath();
      ctx.moveTo(0, -64);
      ctx.lineTo(54, 0);
      ctx.lineTo(0, 64);
      ctx.lineTo(-54, 0);
      ctx.closePath();
      ctx.fill();
      break;
    case "bars":
      for (const bx of [-48, -12, 24]) {
        rounded(ctx, bx, -56, 22, 112, 8);
        ctx.fill();
      }
      break;
    case "slash":
      ctx.lineWidth = 22;
      ctx.beginPath();
      ctx.moveTo(-44, 48);
      ctx.lineTo(44, -48);
      ctx.stroke();
      break;
    case "chevron":
      ctx.lineWidth = 18;
      ctx.beginPath();
      ctx.moveTo(-42, -44);
      ctx.lineTo(38, 0);
      ctx.lineTo(-42, 44);
      ctx.stroke();
      break;
    case "capsule":
      rounded(ctx, -28, -64, 56, 128, 28);
      ctx.fill();
      break;
    case "plus":
      ctx.beginPath();
      ctx.moveTo(-18, -60);
      ctx.lineTo(18, -60);
      ctx.lineTo(18, -18);
      ctx.lineTo(60, -18);
      ctx.lineTo(60, 18);
      ctx.lineTo(18, 18);
      ctx.lineTo(18, 60);
      ctx.lineTo(-18, 60);
      ctx.lineTo(-18, 18);
      ctx.lineTo(-60, 18);
      ctx.lineTo(-60, -18);
      ctx.lineTo(-18, -18);
      ctx.closePath();
      ctx.fill();
      break;
    default: {
      const _x: never = kind;
      return _x;
    }
  }
}

/** Brand mark only: three upright bars, middle slightly skewed. No puzzles, rupee, or chips. */
function drawBrandBars(ctx: CanvasRenderingContext2D, cx: number, cy: number): void {
  const barW = 22;
  const barH = 56;
  const gap = 16;
  const rx = 7;
  ctx.fillStyle = TEXT;
  const xs = [cx - barW * 1.5 - gap, cx - barW / 2, cx + barW / 2 + gap];
  xs.forEach((x, i) => {
    ctx.save();
    ctx.translate(x + barW / 2, cy);
    if (i === 1) ctx.rotate((-8 * Math.PI) / 180);
    rounded(ctx, -barW / 2, -barH / 2, barW, barH, rx);
    ctx.fill();
    ctx.restore();
  });
}

function setTracking(ctx: CanvasRenderingContext2D, value: string): void {
  (ctx as CanvasRenderingContext2D & { letterSpacing: string }).letterSpacing = value;
}

function rounded(
  ctx: CanvasRenderingContext2D,
  x: number,
  y: number,
  w: number,
  h: number,
  r: number,
): void {
  const rr = Math.min(r, w / 2, h / 2);
  ctx.beginPath();
  ctx.moveTo(x + rr, y);
  ctx.arcTo(x + w, y, x + w, y + h, rr);
  ctx.arcTo(x + w, y + h, x, y + h, rr);
  ctx.arcTo(x, y + h, x, y, rr);
  ctx.arcTo(x, y, x + w, y, rr);
  ctx.closePath();
}

export async function shareResult(opts: {
  score: number;
  roundLabel: string;
  newBest: boolean;
  mode: Mode;
  dateId: string;
}): Promise<"shared" | "downloaded" | "copied"> {
  const blob = await renderScoreCard(opts);
  const file = new File([blob], "skew-score.png", { type: "image/png" });
  const text = challengeText(opts.score, location.origin, opts.dateId);
  const nav = navigator as Navigator & {
    canShare?: (data: ShareData) => boolean;
  };
  if (nav.share) {
    try {
      const data: ShareData = { title: "SKEW", text, files: [file] };
      if (!nav.canShare || nav.canShare(data)) {
        await nav.share(data);
        return "shared";
      }
      await nav.share({ title: "SKEW", text });
      downloadBlob(blob, file.name);
      return "shared";
    } catch (err) {
      if ((err as DOMException).name === "AbortError") return "shared";
    }
  }
  downloadBlob(blob, file.name);
  try {
    await navigator.clipboard.writeText(text);
    return "copied";
  } catch {
    return "downloaded";
  }
}

function downloadBlob(blob: Blob, name: string): void {
  const url = URL.createObjectURL(blob);
  const a = document.createElement("a");
  a.href = url;
  a.download = name;
  a.click();
  setTimeout(() => URL.revokeObjectURL(url), 4000);
}

export function isUpiCapable(): boolean {
  return /Android|iPhone|iPad|iPod/i.test(navigator.userAgent);
}

File: android/app/src/main/java/app/skew/android/share/Share.kt
package app.skew.android.share

import android.content.ClipData
import android.content.Context
import android.content.Intent
import android.graphics.Bitmap
import android.graphics.Canvas
import android.graphics.Paint
import android.graphics.Path
import android.graphics.RectF
import android.graphics.Typeface
import android.net.Uri
import androidx.core.content.FileProvider
import app.skew.android.Config
import app.skew.android.game.Mode
import app.skew.android.game.Rng
import app.skew.android.game.dailyPath
import app.skew.android.game.formatScore
import app.skew.android.game.hashString
import java.io.File
import java.io.FileOutputStream

enum class ShareSil { CIRCLE, TRIANGLE, DIAMOND, BARS, SLASH, CHEVRON, CAPSULE, PLUS }

private val SHARE_SILS = listOf(
    ShareSil.CIRCLE, ShareSil.TRIANGLE, ShareSil.DIAMOND, ShareSil.BARS,
    ShareSil.SLASH, ShareSil.CHEVRON, ShareSil.CAPSULE, ShareSil.PLUS,
)

fun shareLockup(siteUrl: String = Config.SITE_URL): String {
    val raw = siteUrl.trimEnd('/')
    return raw.removePrefix("https://").removePrefix("http://").substringBefore('/')
}

fun challengeOrigin(fallback: String = Config.SITE_URL): String =
    (Config.SITE_URL.ifBlank { fallback }).trimEnd('/')

fun challengeText(score: Int, dateId: String, origin: String = Config.SITE_URL): String {
    val link = "${challengeOrigin(origin)}${dailyPath(dateId)}"
    return "SKEW — Find the fake. I scored ${formatScore(score)}. Can you beat me?\n$link"
}

fun shareSilhouetteKey(mode: Mode, dateId: String): String =
    if (mode == Mode.DAILY) "daily:$dateId" else "endless"

fun shareSilhouettes(mode: Mode, dateId: String): List<ShareSil> {
    val seed = hashString("${Config.SHARE_SIL_VERSION}:${shareSilhouetteKey(mode, dateId)}")
    val rng = Rng(seed)
    return listOf(rng.pick(SHARE_SILS), rng.pick(SHARE_SILS), rng.pick(SHARE_SILS))
}

fun cardDateLine(mode: Mode, dateId: String): String? =
    if (mode == Mode.DAILY && dateId.isNotEmpty()) "$dateId UTC" else null

fun cardMetaLine(roundLabel: String, newBest: Boolean): String =
    if (newBest) "$roundLabel  ·  New best" else roundLabel

fun renderScoreCard(
    score: Int,
    roundLabel: String,
    newBest: Boolean,
    dateId: String,
    mode: Mode,
): Bitmap {
    val w = 1080
    val h = 1350
    val bmp = Bitmap.createBitmap(w, h, Bitmap.Config.ARGB_8888)
    val c = Canvas(bmp)
    c.drawColor(0xFF0B0B0C.toInt())
    val text = Paint(Paint.ANTI_ALIAS_FLAG).apply {
        color = 0xFFF4F1EA.toInt()
        textAlign = Paint.Align.CENTER
        typeface = Typeface.create(Typeface.SANS_SERIF, Typeface.BOLD)
        textSize = 64f
        letterSpacing = 0.16f
    }
    c.drawText("SKEW", w / 2f, 150f, text)
    drawBrandBars(c, w / 2f, 230f)
    drawSilBoard(c, w / 2f, 520f, shareSilhouettes(mode, dateId))
    text.letterSpacing = 0f
    text.textSize = 128f
    c.drawText(formatScore(score), w / 2f, 820f, text)
    text.textSize = 34f
    text.typeface = Typeface.create(Typeface.MONOSPACE, Typeface.NORMAL)
    c.drawText(cardMetaLine(roundLabel, newBest), w / 2f, 880f, text)
    val dateLine = cardDateLine(mode, dateId)
    if (dateLine != null) {
        text.color = 0xFF8A8680.toInt()
        text.textSize = 28f
        c.drawText(dateLine, w / 2f, 928f, text)
    }
    text.typeface = Typeface.create(Typeface.SANS_SERIF, Typeface.NORMAL)
    text.color = 0xFF8A8680.toInt()
    text.textSize = 34f
    c.drawText("Can you beat me?", w / 2f, 1040f, text)
    val accent = Paint(Paint.ANTI_ALIAS_FLAG).apply { color = 0xFFE8FF47.toInt() }
    c.drawRect(w / 2f - 48f, 1076f, w / 2f + 48f, 1084f, accent)
    text.color = 0xFFF4F1EA.toInt()
    text.textSize = 26f
    text.typeface = Typeface.create(Typeface.SANS_SERIF, Typeface.BOLD)
    text.letterSpacing = 0.02f
    c.drawText(shareLockup(), w / 2f, 1188f, text)
    return bmp
}

private fun drawSilBoard(c: Canvas, cx: Float, cy: Float, sils: List<ShareSil>) {
    val cardW = 220f
    val cardH = 300f
    val gap = 28f
    val total = cardW * 3 + gap * 2
    var x = cx - total / 2
    val surface = Paint(Paint.ANTI_ALIAS_FLAG).apply { color = 0xFF161618.toInt() }
    val glyph = Paint(Paint.ANTI_ALIAS_FLAG).apply {
        color = 0xFFF4F1EA.toInt()
        strokeCap = Paint.Cap.ROUND
        strokeJoin = Paint.Join.ROUND
        style = Paint.Style.FILL
    }
    for (kind in sils) {
        c.drawRoundRect(RectF(x, cy - cardH / 2, x + cardW, cy + cardH / 2), 28f, 28f, surface)
        c.save()
        c.translate(x + cardW / 2, cy)
        drawSil(c, kind, glyph)
        c.restore()
        x += cardW + gap
    }
}

private fun drawSil(c: Canvas, kind: ShareSil, paint: Paint) {
    val fill = Paint(paint).apply { style = Paint.Style.FILL }
    val stroke = Paint(paint).apply {
        style = Paint.Style.STROKE
        strokeCap = Paint.Cap.ROUND
        strokeJoin = Paint.Join.ROUND
    }
    when (kind) {
        ShareSil.CIRCLE -> c.drawCircle(0f, 0f, 58f, fill)
        ShareSil.TRIANGLE -> {
            val p = Path()
            p.moveTo(0f, -64f)
            p.lineTo(58f, 52f)
            p.lineTo(-58f, 52f)
            p.close()
            c.drawPath(p, fill)
        }
        ShareSil.DIAMOND -> {
            val p = Path()
            p.moveTo(0f, -64f)
            p.lineTo(54f, 0f)
            p.lineTo(0f, 64f)
            p.lineTo(-54f, 0f)
            p.close()
            c.drawPath(p, fill)
        }
        ShareSil.BARS -> {
            for (bx in floatArrayOf(-48f, -12f, 24f)) {
                c.drawRoundRect(RectF(bx, -56f, bx + 22f, 56f), 8f, 8f, fill)
            }
        }
        ShareSil.SLASH -> {
            stroke.strokeWidth = 22f
            c.drawLine(-44f, 48f, 44f, -48f, stroke)
        }
        ShareSil.CHEVRON -> {
            stroke.strokeWidth = 18f
            val p = Path()
            p.moveTo(-42f, -44f)
            p.lineTo(38f, 0f)
            p.lineTo(-42f, 44f)
            c.drawPath(p, stroke)
        }
        ShareSil.CAPSULE -> c.drawRoundRect(RectF(-28f, -64f, 28f, 64f), 28f, 28f, fill)
        ShareSil.PLUS -> {
            val p = Path()
            p.moveTo(-18f, -60f)
            p.lineTo(18f, -60f)
            p.lineTo(18f, -18f)
            p.lineTo(60f, -18f)
            p.lineTo(60f, 18f)
            p.lineTo(18f, 18f)
            p.lineTo(18f, 60f)
            p.lineTo(-18f, 60f)
            p.lineTo(-18f, 18f)
            p.lineTo(-60f, 18f)
            p.lineTo(-60f, -18f)
            p.lineTo(-18f, -18f)
            p.close()
            c.drawPath(p, fill)
        }
    }
}

private fun drawBrandBars(c: Canvas, cx: Float, cy: Float) {
    val barW = 22f
    val barH = 56f
    val gap = 16f
    val paint = Paint(Paint.ANTI_ALIAS_FLAG).apply { color = 0xFFF4F1EA.toInt() }
    val xs = floatArrayOf(cx - barW * 1.5f - gap, cx - barW / 2f, cx + barW / 2f + gap)
    xs.forEachIndexed { i, x ->
        c.save()
        c.translate(x + barW / 2f, cy)
        if (i == 1) c.rotate(-8f)
        c.drawRoundRect(RectF(-barW / 2f, -barH / 2f, barW / 2f, barH / 2f), 7f, 7f, paint)
        c.restore()
    }
}

fun shareScore(
    context: Context,
    score: Int,
    roundLabel: String,
    newBest: Boolean,
    dateId: String,
    mode: Mode,
) {
    val text = challengeText(score, dateId)
    val bmp = renderScoreCard(score, roundLabel, newBest, dateId, mode)
    val dir = File(context.cacheDir, "share").apply { mkdirs() }
    val file = File(dir, "skew-score.png")
    FileOutputStream(file).use { bmp.compress(Bitmap.CompressFormat.PNG, 100, it) }
    val uri: Uri = FileProvider.getUriForFile(context, "${context.packageName}.fileprovider", file)
    val intent = Intent(Intent.ACTION_SEND).apply {
        type = "image/png"
        putExtra(Intent.EXTRA_TEXT, text)
        putExtra(Intent.EXTRA_STREAM, uri)
        clipData = ClipData.newUri(context.contentResolver, "SKEW", uri)
        addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION)
    }
    context.startActivity(Intent.createChooser(intent, "SKEW"))
}

fun shareChallenge(context: Context, score: Int, dateId: String) {
    val text = challengeText(score, dateId)
    val intent = Intent(Intent.ACTION_SEND).apply {
        type = "text/plain"
        putExtra(Intent.EXTRA_TEXT, text)
        putExtra(Intent.EXTRA_SUBJECT, "SKEW")
    }
    context.startActivity(Intent.createChooser(intent, "SKEW"))
}
